### ISSUE TITLE
8.0 RC 1 regression:  tartree does not work for .tar.gz files

### ISSUE THREAD

**Issue by cocorossello** [gradle/gradle#23391]
Tartree seems to be broken in 8.0-RC1

The following code works in gradle 7.6 but throws an error in 8.0-RC1. 

```
plugins {
    id "de.undercouch.download" version "5.3.0"
}

tasks.register("download", Download) {
    src("https://github.com/sass/dart-sass/releases/download/1.57.1/dart-sass-1.57.1-linux-x64.tar.gz")
    dest file("$buildDir")
}

tasks.register("unpack", Copy) {
    dependsOn tasks.named("download")
    from tarTree("$buildDir/dart-sass-1.57.1-linux-x64.tar.gz")
    into file("$buildDir/somefolder")
}
```
`./gradlew unpack`

### Expected Behavior
File should be unpacked

```
### Current Behavior

> Task :download
Download https://github.com/sass/dart-sass/releases/download/1.57.1/dart-sass-1.57.1-linux-x64.tar.gz
SLF4J: No SLF4J providers were found.
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See https://www.slf4j.org/codes.html#noProviders for further details.

> Task :unpack FAILED
2 actionable tasks: 2 executed

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':unpack'.
> Unable to expand TAR 'C:\proyectos\gradle-jartree\build\dart-sass-1.57.1-linux-x64.tar.gz'
    The tar might be corrupted or it is compressed in an unexpected way.
    By default the tar tree tries to guess the compression based on the file extension.
    If you need to specify the compression explicitly please refer to the DSL reference.

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.

```

### Steps to Reproduce 
I provided is a minimal reproducer

### Your Environment
Tested in windows and linux, in java 17 and 18
Build scan URL:
https://gradle.com/s/gipau4zcwhi3e

**Comment by eskatos** [comment]
Thank you for providing a valid reproducer.

The regression will be taken care of.

**Comment by eskatos** [comment]
* Fixed in https://github.com/gradle/gradle/pull/23402

**Comment by mkjois** [comment]
I ran into this same bug recently, and I wanted to post my resolution in case anyone comes across it in the future, and I haven't seen this documented as a possible cause.

So my use case is nearly the same as the one in this post, except I'm downloading the [Boost source code](https://www.boost.org/users/download/). The URL is this:
```
https://archives.boost.io/release/1.84.0/source/boost_1_84_0.tar.gz
```

My environment is a stock `amazonlinux:2023` Docker container. Actually I'm also building on Mac OS X but Gradle seems to handle the tarball correctly there.

There is nothing wrong with the decompression, whether it's `gz` or `xz` or `bz2`. The error happens once it gets decompressed into `.tar` format. What I found after digging through Gradle code and Apache `commons-compress` code, is that file names in each tar entry get decoded with Java's default `Charset`, which is based on the _system's locale_, and something in the Boost tarball doesn't decode properly when that's ASCII. The fix is to make sure the system locale is some form of UTF-8.

The Gradle stacktrace led me to look through the following code:
- Gradle [`TarFileTree.visitImpl()`](https://github.com/gradle/gradle/blob/v8.5.0/subprojects/core/src/main/java/org/gradle/api/internal/file/archive/TarFileTree.java#L112)
- Commons Compress [`TarArchiveEntry.parseTarHeader()`](https://github.com/apache/commons-compress/blob/rel/1.21/src/main/java/org/apache/commons/compress/archivers/tar/TarArchiveEntry.java#L1565)
- Commons Compress [`TarUtils.DEFAULT_ENCODING`](https://github.com/apache/commons-compress/blob/rel/1.21/src/main/java/org/apache/commons/compress/archivers/tar/TarUtils.java#L53)
- Commons Compress [`ZipEncodingHelper.getZipEncoding()`](https://github.com/apache/commons-compress/blob/rel/1.21/src/main/java/org/apache/commons/compress/archivers/zip/ZipEncodingHelper.java#L54)

Here's a tiny Java program you can compile and run to test what `commons-compress` will use for its default Charset:
```java
public class SystemCharsetTest {
    public static void main(final String[] args) {
        System.out.println(java.nio.charset.Charset.defaultCharset());
    }
}
```

On my Mac Sonoma 14.2.1, this prints `UTF-8` which is why Gradle makes it work out of the box. But on the stock Linux Docker container I'm using, it prints `US-ASCII`. Setting the locale like this made it work:
```sh
export LC_ALL=C.UTF8
```

It doesn't have to be specifically that; run `locale -a` and pick any that ends with `UTF-8`. Often times you'll see something like `en_US.UTF-8` which is pretty standard. You can even find ways to install extra regional UTF-8 locales if needed.

### TL;DR

Make sure your system locale is some form of UTF-8, especially if you're in a Docker container.
